pkg/cloudprovider/aws/sqs.go

```diff
@@ -0,0 +1,258 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/sqs"
+	"github.com/aws/aws-sdk-go/service/sqs/sqsiface"
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/utils/functional"
+	"github.com/aws/karpenter/pkg/utils/injection"
+)
+
+type QueuePolicy struct {
+	Version   string                 `json:"Version"`
+	ID        string                 `json:"Id"`
+	Statement []QueuePolicyStatement `json:"Statement"`
+}
+
+type QueuePolicyStatement struct {
+	Effect    string    `json:"Effect"`
+	Principal Principal `json:"Principal"`
+	Action    []string  `json:"Action"`
+	Resource  string    `json:"Resource"`
+}
+
+type Principal struct {
+	Service []string `json:"Service"`
+}
+
+type SQSProvider struct {
+	client sqsiface.SQSAPI
+
+	createQueueInput    *sqs.CreateQueueInput
+	getQueueURLInput    *sqs.GetQueueUrlInput
+	receiveMessageInput *sqs.ReceiveMessageInput
+	mutex               *sync.RWMutex
+	queueURL            string
+	queueName           string
+	metadata            *Metadata
+}
+
+func NewSQSProvider(ctx context.Context, client sqsiface.SQSAPI, metadata *Metadata) *SQSProvider {
+	provider := &SQSProvider{
+		client:    client,
+		mutex:     &sync.RWMutex{},
+		metadata:  metadata,
+		queueName: getQueueName(ctx),
+	}
+	provider.createQueueInput = &sqs.CreateQueueInput{
+		Attributes: provider.getQueueAttributes(),
+		QueueName:  aws.String(provider.queueName),
+		Tags: map[string]*string{
+			v1alpha5.DiscoveryLabelKey: aws.String(injection.GetOptions(ctx).ClusterName),
+		},
+	}
+	provider.getQueueURLInput = &sqs.GetQueueUrlInput{
+		QueueName: aws.String(provider.queueName),
+	}
+	provider.receiveMessageInput = &sqs.ReceiveMessageInput{
+		MaxNumberOfMessages: aws.Int64(10),
+		VisibilityTimeout:   aws.Int64(10), // Seconds
+		WaitTimeSeconds:     aws.Int64(10), // Seconds, maximum for long polling
```

Comment:
I guess I'll ask the same question back. What was the reason for the original 20 second heuristic?

Reply:
Discussed offline, NTH used 20 to reduce network calls since there was no real downside to maxing out the wait period that I know of at least.  

Also I wasn't sure which method did a full query when we chatted, but looks like that is the long polling method:

```
With long polling, the ReceiveMessage request queries all of the servers for messages. Amazon SQS sends a response after it collects at least one available message, up to the maximum number of messages specified in the request. Amazon SQS sends an empty response only if the polling wait time expires.
```

https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-short-and-long-polling.html